src/frontend/src/expr/function_call.rs

```diff
@@ -175,6 +175,77 @@ impl FunctionCall {
                 })
             }
+            ExprType::ArrayCat => {
+                if inputs.len() != 2 {
+                    return Err(ErrorCode::BindError(format!(
+                        "Function `ArrayCat` takes 2 arguments ({} given)",
+                        inputs.len()
+                    ))
+                    .into());
+                }
+                let left_type = inputs[0].return_type();
+                let right_type = inputs[1].return_type();
+
+                let wrap_single_elem = |elem, ret_type| -> ExprImpl {
+                    Self {
+                        func_type: ExprType::Array,
+                        return_type: ret_type,
+                        inputs: vec![elem],
+                    }
+                    .into()
+                };
+
+                let return_type = match (&left_type, &right_type) {
+                    (
+                        DataType::List {
+                            datatype: left_elem_type,
+                        },
+                        DataType::List {
+                            datatype: right_elem_type,
+                        },
+                    ) => {
+                        if **left_elem_type == **right_elem_type {
+                            Some(left_type.clone())
+                        } else if **left_elem_type == right_type {
+                            inputs[1] = wrap_single_elem(inputs[1].clone(), left_type.clone());
+                            Some(left_type.clone())
+                        } else if left_type == **right_elem_type {
+                            inputs[0] = wrap_single_elem(inputs[0].clone(), right_type.clone());
+                            Some(right_type.clone())
+                        } else {
+                            None
+                        }
+                    }
+                    (
+                        DataType::List {
+                            datatype: left_elem_type,
+                        },
+                        _,
+                    ) if **left_elem_type == right_type => {
+                        inputs[1] = wrap_single_elem(inputs[1].clone(), left_type.clone());
+                        Some(left_type.clone())
+                    }
+                    (
+                        _,
+                        DataType::List {
+                            datatype: right_elem_type,
+                        },
+                    ) if left_type == **right_elem_type => {
+                        inputs[0] = wrap_single_elem(inputs[0].clone(), right_type.clone());
+                        Some(right_type.clone())
+                    }
+                    _ => None,
+                };
+                if let Some(return_type) = return_type {
+                    Ok(return_type)
+                } else {
+                    return Err(ErrorCode::BindError(format!(
+                        "Wrong argument types for function `ArrayCat` ({:?} and {:?})",
```

Comment:
```suggestion
                        "Wrong argument types for function `ArrayCat` ({} and {})",
```

```sql
-- before
dev=> select array_cat(Array[1], Array[Array[Array[3]]]);
ERROR:  QueryError: Bind error: Wrong argument types for function `ArrayCat` (List { datatype: Int32 } and List { datatype: List { datatype: List { datatype: Int32 } } })

-- after
dev=> select array_cat(Array[1], Array[Array[Array[3]]]);
ERROR:  QueryError: Bind error: Wrong argument types for function `ArrayCat` (integer[] and integer[][][])
```

BTW, it might be better to also include error tests. (P.S. unfortunately slt doesn't have full error support yet 🤪 https://github.com/risinglightdb/sqllogictest-rs/issues/30. 
Currently this is ok:
```
statement error
select array_cat(Array[1], Array[Array[Array[3]]]);
```
P.P.S alternatively the error can also be tested in planner test...

Reply:
+1 for planner test.